base/rational_to_float.jl

```diff
@@ -0,0 +1,246 @@
+        (false, false)
+    else
+        let from = Rnd.rounds_away_from_zero(rm, sb)
+            (!from, from)
+        end
+    end::NTuple{2,Bool}
+    exp_max = Base.exponent_max(T)
+    exp_min = Base.exponent_min(T)
+    ieee_repr = Base.ieee754_representation
+    repr_zero = ieee_repr(T, sb, Val(:zero))
+    repr_inf  = ieee_repr(T, sb, Val(:inf))
+    repr_nan  = ieee_repr(T, sb, Val(:nan))
+    U = typeof(repr_zero)
+    repr_zero::U
+    repr_inf::U
+    repr_nan::U
+
+    ret_u = if is_regular
+        let
+            c = let e = exp_min - 1
+                rational_to_float_components(S, num, den, prec, e, rm, sb)
+            end
+            exp = c.exponent
+            exp_diff = exp - exp_min
+            is_normal = 0 ≤ exp_diff
+            exp_is_huge_p = exp_max < exp
+            exp_is_huge_n = signbit(exp_diff + prec)
+            rounds_to_inf  = exp_is_huge_p & !rm_is_to_zero
+            rounds_to_zero = exp_is_huge_n & !rm_is_from_zero
+
+            if !rounds_to_zero & !exp_is_huge_p
+                let signif = (c.integral_significand % U) & Base.significand_mask(T)
+                    exp_field = (max(exp_diff, zero(exp_diff)) + is_normal) % U
+                    ieee_repr(T, sb, exp_field, signif)::U
+                end
+            elseif rounds_to_zero
+                repr_zero
+            elseif rounds_to_inf
+                repr_inf
+            else
+                ieee_repr(T, sb, Val(:omega))
+            end
+        end
+    else
+        if is_zero
+            repr_zero
+        elseif is_inf
+            repr_inf
+        else
+            repr_nan
+        end
+    end::U
+
+    reinterpret(T, ret_u)::T
+end
+
+# `BigInt` is a safe default.
+rational_to_float_promote_type(::Type{F}, ::Type{S}) where {F,S} = BigInt
+
+const BitUnsignedOrBool = Union{Bool,Base.BitUnsigned}
+const BitIntegerOrBool  = Union{Bool,Base.BitInteger}
+const BitSigned = Base.BitSigned
+
+integer_type_with_size(::Type{<:BitUnsignedOrBool}, ::Val{1})  = UInt16
+integer_type_with_size(::Type{<:BitUnsignedOrBool}, ::Val{2})  = UInt16
+integer_type_with_size(::Type{<:BitUnsignedOrBool}, ::Val{4})  = UInt32
+integer_type_with_size(::Type{<:BitUnsignedOrBool}, ::Val{8})  = UInt64
+integer_type_with_size(::Type{<:BitUnsignedOrBool}, ::Val{16}) = UInt128
+integer_type_with_size(::Type{<:BitSigned},         ::Val{1})  =  Int16
+integer_type_with_size(::Type{<:BitSigned},         ::Val{2})  =  Int16
+integer_type_with_size(::Type{<:BitSigned},         ::Val{4})  =  Int32
+integer_type_with_size(::Type{<:BitSigned},         ::Val{8})  =  Int64
+integer_type_with_size(::Type{<:BitSigned},         ::Val{16}) =  Int128
+
+# As an optimization, use an integer type narrower than `BigInt` when possible.
+function rational_to_float_promote_type(::Type{F}, ::Type{S}) where {F<:Base.IEEEFloat,S<:BitIntegerOrBool}
+    m = max(sizeof(F), sizeof(S))::Int
+    Max = integer_type_with_size(S, Val(m))
+    widen(Max)
```

Comment:
```suggestion
    if sizeof(S) > sizeof(F)
        widen(S)
    else
        S <: Signed ? widen(Base.inttype(F)) : widen(Base.uinttype(F))
    end
```

You can then remove `integer_type_with_size`.